### ISSUE TITLE
Allow to annotate pods with cni.projectcalico.org/floatingIPs dynamically

### ISSUE THREAD

**Issue by hightoxicity** [projectcalico/cni-plugin#763]
Hi,

I have some pods which need to egress (for example recursive dns), each of those pod should pick one IP from a pool in a consistent way.
To achieve this goal, I decided to use a statefulset with an array of public ips...
Using the statefulset replica index, I choose the same IP each time for a replica but now I need to upgrade the pods annotation (cni.projectcalico.org/floatingIPs) to make calico announce the picked IP from the pool.

It currently does not work, it is only working on pod creation having the annotation embbeded into the definition.

Looking at how it is currently working, it seems hard to use replica feature of deployment/statefulset with calico floating ip feature...

**Comment by caseydavenport** [comment]
@hightoxicity yeah, that's right - at the moment the floatingIPs annotation is only respected on pod creation.

I think if we wanted to make it adjustable at runtime, we'd need to add code in the Calico pod controller similar to what we do for labels. 

This is the bit of code that would need to be changed: https://github.com/projectcalico/kube-controllers/blob/master/pkg/controllers/pod/pod_controller.go

**Comment by hightoxicity** [comment]
Thx for your response, I think this would be a killer feature :-)
Today I am doing some helm ranges logic to generate one deployment by replica for each outgoing IP as a workaround.
I am quite surprised nobody asked something about it yet, it concerns mainly people working with bare metal servers directly and people working in an openstack context for example (maybe they are all setting a default route to NAT instances, something I do not want)...

**Comment by caseydavenport** [comment]
Agree this would be nice. I'd be happy to help with design and code review for anyone who wants to give this a try!

**Comment by hightoxicity** [comment]
:-) I will try to convince my company that I should do something for this!
Great thanks.

**Comment by hprateek43** [comment]
I can work on this issue. Will need a context around it though. @caseydavenport Can you provide some context around this?

**Comment by caseydavenport** [comment]
@hprateek43 right now this is only configurable at pod-launch time, since it is handled in the CNI plugin which only gets executed when the pod is starting up.

We want to enhance the pod controller in Calico's `kube-controllers` so that it respects this annotation, and updates the Calico workload endpoint with a floating IP even if it changes after the pod is running. 

See this comment for a hint on where that code lives: https://github.com/projectcalico/cni-plugin/issues/763#issuecomment-510253529

**Comment by hprateek43** [comment]
@caseydavenport I will be following up on this. Does calico have a slack where can collaborate, or are you active on Kubernetes slack?

**Comment by tmjd** [comment]
You can get a slack invite at https://www.projectcalico.org/community.

**Comment by ashish-billore** [comment]
This sounds interesting and quite useful feature of dynamic movement of floating IP.
@caseydavenport  @hprateek43 any plans for this being implemented in Calico?

**Comment by hprateek43** [comment]
This change requires changes in upstream repos as well. Not sure if we are pursuing it. I worked on this but then it was identified that changes in other packages are required. @caseydavenport Is this a part of some milestone ?

**Comment by caseydavenport** [comment]
I'd love to see this implemented, and would be glad to help. I don't think any of the core Calico team are planning on working on this in the immediate future, though.

Reach out here or on Slack if you're interested in working on this. I think most of the code is in https://github.com/projectcalico/libcalico-go and https://github.com/projectcalico/kube-controllers
